### PR TITLE
[GEOS-9322] WFS-NG remote re-projection

### PR DESCRIPTION
**Pull request by imranrajjad**
when 'usedefaultsrs' is set to false in WFS DataStore, remote requests should be relayed in requested SRS if they exist as supported SRS for native remote feature to skip re-projection.

<Include a few sentences describing the overall goals for this Pull Request>

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for community modules):
- [x] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [x] New unit tests have been added covering the changes
- [ x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by travis-ci after opening this PR)
- [x] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.

### CODE REVIEW

**Comment by imranrajjad** [comment]
https://osgeo-org.atlassian.net/browse/GEOS-9322

**Comment by imranrajjad** [comment]
@nmco @aaime would be glad if this PR can be reviewed

File: src/main/src/main/java/org/geoserver/catalog/FeatureTypeInfo.java [hunk 1]
```diff
@@ -30,6 +30,9 @@
     /** The cascaded stored query configuration */
     static final String STORED_QUERY_CONFIGURATION = "WFS_NG_STORED_QUERY_CONFIGURATION";
 
+    /** ONLY FOR WFS-NG Feature Types */
```
Comment (nmco) [inline src/main/src/main/java/org/geoserver/catalog/FeatureTypeInfo.java]: If it is only for ``WFS-NG Feature Types`` (not sue what does this means) the it shoudl not be here.

File: src/main/src/main/java/org/geoserver/catalog/ResourcePool.java [hunk 2]
```diff
@@ -5,7 +5,7 @@
  */
 package org.geoserver.catalog;
 
-import java.awt.*;
+import java.awt.RenderingHints;
```
Comment (nmco) [inline src/main/src/main/java/org/geoserver/catalog/ResourcePool.java]: ?

**Comment by nmco** [comment]
Associated issue:
https://osgeo-org.atlassian.net/browse/GEOS-9322

@imranrajjad can you write here a small summary explaining:
- what's wrong with ``usedefaultsrs`` property, do you foreseen backwards compatibility issues?
- expalining a bit your plan, there is a lot of code changes :smile: 

If you could do that small summary it would help me a lot navigating through the changes.

File: src/main/src/main/java/org/geoserver/catalog/FeatureTypeInfo.java (hunk 1 above)
Comment (imranrajjad) [reply src/main/src/main/java/org/geoserver/catalog/FeatureTypeInfo.java]: this is key that will be used to store other supported SRS for a remote Feature type coming from WFS capability document. If the GetMap or GetFeature request have spatial filter with SRS belonging to one of the stored SRS, then the changes of this PR will ensure that they are passed down to the query level and no un-necessary re-projection occurs.

File: src/main/src/main/java/org/geoserver/catalog/ResourcePool.java (hunk 2 above)
Comment (imranrajjad) [reply src/main/src/main/java/org/geoserver/catalog/ResourcePool.java]: will revert, that's probably eclipse being smart

File: src/main/src/main/java/org/geoserver/catalog/CatalogBuilder.java [hunk 3]
```diff
@@ -382,6 +382,25 @@ public FeatureTypeInfo buildFeatureType(FeatureSource featureSource) {
...
+                                .put(
+                                        FeatureTypeInfo.OTHER_SRS,
+                                        String.join(",", info.getOtherSRS()));
+                }
+            }
+
+        } catch (UnsupportedOperationException ue) {
+            LOGGER.warning("Other SRS not read from Feature Source");
+        }
+
```
Comment (imranrajjad) [inline src/main/src/main/java/org/geoserver/catalog/CatalogBuilder.java]: saving other supported SRS for this feature type if available

File: src/main/src/main/java/org/geoserver/catalog/ResourcePool.java [hunk 4]
```diff
@@ -1308,6 +1309,11 @@ public void clear(FeatureTypeInfo info) {
                 resultCRS = nativeCRS;
             } else {
                 resultCRS = getCRS(info.getSRS());
+                // force remoting re-projection incase of WFS-NG only
+                if (hints != null)
+                    if (hints.get(ResourcePool.MAP_CRS) != null
+                            && info.getMetadata().get(FeatureTypeInfo.OTHER_SRS) != null)
+                        resultCRS = (CoordinateReferenceSystem) hints.get(ResourcePool.MAP_CRS);
             }
```
Comment (imranrajjad) [inline src/main/src/main/java/org/geoserver/catalog/ResourcePool.java]: setting the result CRS to Getmap/GetFeature`s spatial filter, if the CRS is one of the OTHER supported SRS of this featuretype coming from WFS-NG Datastore

**Note to self: Verify the datasrouce is WFS-NG with usedefaultsrs set to false**

File: src/main/src/main/java/org/vfny/geoserver/global/GeoServerFeatureSource.java [hunk 5]
```diff
@@ -210,7 +211,7 @@ protected Query makeDefinitionQuery(Query query, SimpleFeatureType schema) throw
             Query defQuery = new Query(query);
             defQuery.setFilter(filter);
             defQuery.setPropertyNames(propNames);
-
+            defQuery.setCoordinateSystem(query.getCoordinateSystem());
```
Comment (imranrajjad) [inline src/main/src/main/java/org/vfny/geoserver/global/GeoServerFeatureSource.java]: setting the CRS of Query to use Geotools capability of skipping re-projection if CRS is remotely supported

File: src/main/src/main/java/org/vfny/geoserver/global/GeoServerFeatureSource.java [hunk 6]
```diff
@@ -570,7 +572,8 @@ protected Query adaptQuery(Query query, SimpleFeatureType schema) throws IOExcep
         if (newQuery.getCoordinateSystemReproject() != null) {
             newQuery.setCoordinateSystemReproject(null);
         }
-        if (newQuery.getCoordinateSystem() != null) {
+        if (newQuery.getCoordinateSystem() != null
+                && metadata.get(FeatureTypeInfo.OTHER_SRS) == null) {
```
Comment (imranrajjad) [inline src/main/src/main/java/org/vfny/geoserver/global/GeoServerFeatureSource.java]: this method removes CRS from query, modified it to not do incase of WFS-NG feature type

File: src/web/core/src/main/java/org/geoserver/web/data/resource/BasicResourceConfig.java [hunk 7]
```diff
@@ -230,6 +247,43 @@ public boolean getDefaultFormProcessing() {
...
+                    return true;
+                }
+            }
+        } catch (IOException e) {
+            Logger.getGlobal().log(Level.SEVERE, e.getMessage(), e);
+        }
+
+        return false;
+    }
+
```
Comment (imranrajjad) [inline src/web/core/src/main/java/org/geoserver/web/data/resource/BasicResourceConfig.java]: this function is used to read other srs and update metadata map of FeautreTypeInfo, during Resource creation/modification on UI

File: src/wms/src/main/java/org/geoserver/wms/GetMap.java [hunk 8]
```diff
@@ -512,7 +512,7 @@ WebMap executeInternal(
                 //
                 // /////////////////////////////////////////////////////////
                 try {
-                    source = mapLayerInfo.getFeatureSource(true);
+                    source = mapLayerInfo.getFeatureSource(true, request.getCrs());
```
Comment (imranrajjad) [inline src/wms/src/main/java/org/geoserver/wms/GetMap.java]: this is modification keeps track of CRS of GetMap request

File: src/wms/src/main/java/org/geoserver/wms/MapLayerInfo.java [hunk 9]
```diff
@@ -289,6 +301,12 @@ public int getCacheMaxAge() {
 
         Hints hints = new Hints(ResourcePool.REPROJECT, Boolean.valueOf(!skipReproject));
 
+        if (userMapCRSForWFSNG(resource, coordinateReferenceSystem)) {
+            // a hint for wfs-ng featuresource to keep the crs in query
+            // to skip un-necessary re-projection
+            hints.put(ResourcePool.MAP_CRS, coordinateReferenceSystem);
```
Comment (imranrajjad) [inline src/wms/src/main/java/org/geoserver/wms/MapLayerInfo.java]: if maplayer is a WFS-NG layer, look for the GetMap CRS inside the feature type metadata, if found, add it to the hints which will be passed down to the SecureFeatureSource reader.

File: src/wms/src/main/java/org/geoserver/wms/MapLayerInfo.java [hunk 10]
```diff
@@ -337,6 +355,43 @@ public int hashCode() {
...
+                try {
+                    return CRS.equalsIgnoreMetadata(CRS.decode(crs), coordinateReferenceSystem);
+                } catch (Exception e) {
+                    LOGGER.log(Level.SEVERE, e.getMessage(), e);
+                }
+            }
+
+        return false;
+    }
+
```
Comment (imranrajjad) [inline src/wms/src/main/java/org/geoserver/wms/MapLayerInfo.java]: the method that during a GetMap lookups the requested CRS inside FeatureTypeInfo metadata to skip re-projection. Method also verified that the datastore is WFS-NG configured with`usedefaultsrs` set to false.

**Comment by imranrajjad** [comment]
@nmco Allow me to explain the PR

This PR basically addresses the issue of Geoserver always using the first available SRS in WFS document.

Before reviewing the PR some points to keep in mind.

Geotools WFS-NG Datasource has a boolean option `usedefaultsrs` when set to false, will force Geotools to lookup the SRS of Spatial filter inside the WFS Capability document, if that SRS exists inside <OtherSRS>...</OtherSRS>, no re-projection will occur (something which is requirement of the client). To make this work the SRS of filter must be set inside the Query.

In Geoserver  `usedefaultsrs` option is available on the Datasource page but is not effective and nor used because Geoserver has its own Projection policy and sticks to Native and Declared SRS. The code changes are the modifications that make Geoserver react to the `usedefaultsrs` configuration of WFS-NG Datastore while keeping the backward compatibility intact. 

I have put comments in the code, please let me know if you need further clarification. I admit this was not a subtle change. Reviewing it is even more challenging.

File: src/web/core/src/main/java/org/geoserver/web/data/resource/BasicResourceConfig.java [hunk 11]
```diff
@@ -230,6 +247,43 @@ public boolean getDefaultFormProcessing() {
         };
     }
 
+    public boolean addOtherSRS(ResourceInfo resourceInfo) {
```
Comment (imranrajjad) [inline src/web/core/src/main/java/org/geoserver/web/data/resource/BasicResourceConfig.java]: this function is used to read other srs and update metadata map of FeautreTypeInfo, during Resource creation/modification on UI

File: src/web/core/src/main/java/org/geoserver/web/data/resource/BasicResourceConfig.java [hunk 12]
```diff
@@ -120,6 +125,18 @@ public BasicResourceConfig(String id, IModel<ResourceInfo> model) {
         refForm.add(projectionPolicy);
 
         refForm.add(new ReprojectionIsPossibleValidator(nativeCRS, declaredCRS, projectionPolicy));
+
+        // for Feature types init empty
+        Label otherSRS = new Label("otherSRS", "");
+        if (addOtherSRS(model.getObject())) {
```
Comment (imranrajjad) [inline src/web/core/src/main/java/org/geoserver/web/data/resource/BasicResourceConfig.java]: adding label on UI about other remotely supported SRS

File: src/wms/src/main/java/org/geoserver/wms/MapLayerInfo.java [hunk 13]
```diff
@@ -337,6 +355,43 @@ public int hashCode() {
         return result;
     }
 
+    private boolean userMapCRSForWFSNG(
```
Comment (imranrajjad) [inline src/wms/src/main/java/org/geoserver/wms/MapLayerInfo.java]: the method that during a GetMap lookups the requested CRS inside FeatureTypeInfo metadata to skip re-projection. Method also verified that the datastore is WFS-NG configured withusedefaultsrs set to false.

**Comment by nmco** [comment]
Thx for the summary and fixes @imranrajjad, looks good :+1:.

**Comment by nmco** [comment]
Looks like unrelated build failures.

**Pull request merged**